### DISCUSSION TITLE
Missing `_` in userGroup regex

### DISCUSSION THREAD

**Discussion by rekup** [remarkjs/remark#853] [Q&A] [answered]
Hello everyone!

Before opening an issue I wanted to ask here if this really is bug. We recently discovered the following error message in our [renovatebot](https://github.com/renovatebot/renovate) logs:

```
 WARN: Error linkifying (repository=foo_bar/baz)
       "err": {
         "message": "Missing `repository` field in `options`",
         "stack": "Error: Missing `repository` field in `options`\n    at Function.github (/usr/src/app/node_modules/remark-github/index.js:99:11)\n    at freeze (/usr/src/app/node_modules/unified/index.js:125:31)\n    at Function.process (/usr/src/app/node_modules/unified/index.js:349:5)\n    at Object.linkify (/usr/src/app/node_modules/renovate/lib/util/markdown.ts:37:6)\n    at getReleaseNotes (/usr/src/app/node_modules/renovate/lib/workers/pr/changelog/release-notes.ts:123:39)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at Object.addReleaseNotes (/usr/src/app/node_modules/renovate/lib/workers/pr/changelog/release-notes.ts:368:24)\n    at Object.getChangeLogJSON (/usr/src/app/node_modules/renovate/lib/workers/pr/changelog/source-gitlab.ts:140:9)\n    at Object.getChangeLogJSON (/usr/src/app/node_modules/renovate/lib/workers/pr/changelog/index.ts:35:13)\n    at embedChangelog (/usr/src/app/node_modules/renovate/lib/workers/repository/changelog/index.ts:7:21)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
       },
 ```

We debugged this issue and apparently the problem is the underscore (`_`) character in the repo username (`foo_bar` in this case). I was able to reproduce the error using the following code:

```js
import {readSync} from 'to-vfile'
import {remark} from 'remark'
import remarkGithub from 'remark-github'

const file = readSync('example.md')

remark()
  .use(remarkGithub, { repository: 'foo_bar/bar' })
  .process(file)
  .then((file) => {
    console.log(String(file))
  })
```

If I add the `_` character to the [userGroup regex](https://github.com/remarkjs/remark-github/blob/main/index.js#L56) in `remark-github` the error message disappears but in the rendered markdown the `_` is escaped:

```
...
*   Issue or PR (`#`): [#1](https://github.com/foo\_bar/bar/issues/1)
*   Issue or PR (`GH-`): [GH-1](https://github.com/foo\_bar/bar/issues/1)
...
```

We are using a selfhosted Gitlab CE instance (the full url to the repo used would be something like `https://gitlab.example.com/foo_bar/baz`.

So my questions are:
1. Is it a bug that the user / group name is not allowed to contain `_` (apparently for github.com underscores are "reserved" for "Enterprise Managed Users" [1] but in gitlab everyone can use them)?
2. Why is the underscore escaped in the output?

Thanks!
Reto

[1]: https://github.com/isaacs/github/issues/906#issuecomment-891674387        "GitHub"

**Comment by wooorm** [comment, marked as answer]
Hi!

- `remark-github` is for GitHub. It does not match GitLab features. That’s not a goal of this project. There would probably be much more work needed to support GitLab flavored markdown (and it’s specifically called `remark-github`, not `remark-gitlab`).
- Do you have an example of a GitHub user with `_` in their name? That thread looks a bit vague.
- Underscores are escaped because they can form emphasis in many very complex cases, such as https://github.com/syntax-tree/mdast-util-to-markdown/issues/25

**Reply by rekup** [reply]
Thank you for taking the time to answer my question.

1. That makes sense. I just saw that renovatebot uses `remark-github` even if the repo is not hosted on GitHub so that's why I asked.
2. Unfortunately I don't and I also do not have access to GitHub Enterprise to verify this.
3. Thanks for the explanation.

Thanks again!
Reto

**Reply by wooorm** [reply]
So you are using gitlab and renovate is using this plugin to generate PR body's for you? Doesn't that mean that the URLs are wrong? 🤔
If the difference between gitlab and github is relatively small, we could add some support for it. But I'm guessing that it's quite a lot of changes?

**Reply by rekup** [reply]
Yes, we are using gitlab and renovatebot is creating PRs using repos hosted on the same gitlab instance. Our current releases do not contain much detail so I guess that's why we haven't had any wrong URLs yet (I need to do further testing to verify this tough). I also found that renovatebot is skipping this step if the repo is hosted on gitlab.com: https://github.com/renovatebot/renovate/blob/main/lib/workers/pr/changelog/release-notes.ts#L117. Obviously this does not work  for self hosted gitlab instances.

Currently I can't tell if the differences between github and gitlab are small or not, I'd have to do some more testing here as well (JavaScript is not really my strength to be honest)

**Reply by wooorm** [reply]
You might perhaps be better of raising this with renovate? That they’d do something smarter than `baseUrl !== 'https://gitlab.com/'`?

**Reply by rekup** [reply]
Yes definitely I will do so but first I will have to understand the renovate code a bit better to make sure I've configured everything properly.
Thanks again for your support!
